### PR TITLE
container_push: added attribute 'tag_file'

### PR DESCRIPTION
**Pull request by excavador**
I need ability to specify tag of image for push as content of file.
I added `tag_file` attribute to `container_push` rule.

### CODE REVIEW

File: container/push.bzl [hunk 1]
```diff
@@ -167,6 +171,10 @@ container_push = rule(
             default = "latest",
             doc = "(optional) The tag of the image, default to 'latest'.",
         ),
+        "tag_file": attr.label(
+            allow_single_file = True,
+            doc = "(options) The label of the file with tag value. Overrides 'tag'.",
```
Comment (xingao267) [inline container/push.bzl]: (optional)

File: container/push.bzl [hunk 2]
```diff
@@ -99,6 +99,10 @@ def _impl(ctx):
         "--stamp-info-file=%s" % _get_runfile_path(ctx, f)
         for f in stamp_inputs
     ]
+    runfiles_tag_file = []
```
Comment (xingao267) [inline container/push.bzl]: Move this new section to be after line 90 so it's more clear how tag is set.

**Comment by excavador** [comment]
Will do, thank you!

**Comment by excavador** [comment]
> One that is similar to https://github.com/bazelbuild/rules_docker/blob/master/tests/docker/BUILD#L262 and https://github.com/bazelbuild/rules_docker/blob/master/testing/e2e.sh#L390

Done

> And another file_test (example: https://github.com/GoogleContainerTools/base-images-docker/blob/master/tests/package_managers/BUILD#L32) that validates that the generated executable file has the desired content (e.g. the tag that reads from a file)

Should I add this test to rules_docker or to https://github.com/GoogleContainerTools/base-images-docker/?

**Comment by xingao267** [comment]
In rules_docker as well. Probably in file
https://github.com/bazelbuild/rules_docker/blob/master/tests/docker/BUILD
as well. Similar to other file_test.

**Comment by excavador** [comment]
> In rules_docker as well. Probably in file https://github.com/bazelbuild/rules_docker/blob/master/tests/docker/BUILD as well. Similar to other file_test.

I see. Done.

**Comment by excavador** [comment]
@xingao267 fixed PR comments

**Comment by excavador** [comment]
rebased to latest master

**Comment by excavador** [comment]
@xingao267 everything fixed, tests added, PR rebased to latest master.

File: tests/docker/BUILD [hunk 3]
```diff
@@ -285,6 +304,12 @@ file_test(
     file = ":push_test.digest",
 )
 
+file_test(
+    name = "test_push_tag_file_digest_output",
```
Comment (xingao267) [inline tests/docker/BUILD]: Don't check the digest output, check the default output executable that it contains the right tag.

File: tests/docker/BUILD [hunk 4]
```diff
@@ -269,6 +269,25 @@ container_push(
...
+    name = "push_test_tag_file",
+    format = "Docker",
+    image = ":set_cmd",
+    registry = "localhost:5000",
+    repository = "docker/test",
+    tag_file = ":test.tag",
+    tags = ["manual"],
+)
+
+
```
Comment (xingao267) [inline tests/docker/BUILD]: extra line

**Comment by excavador** [comment]
@xingao267 can you please point me to specific example of the test, where tag checked?
I re-read all rules_docker completely and did not understand what the right way to do it.
I have only single idea: in shell script I should execute bazel run <mytarget>, than try to pull container by tag and compare digests

**Comment by xingao267** [comment]
@excavador So when you build the new container_push target //tests/docker:push_test_tag_file you created, it should create several outputs, one is `bazel-bin/tests/docker/push_test_tag_file.digest` and one is `bazel-bin/tests/docker/push_test_tag_file`. So what I was saying in the comment is do not use the `:push_test_tag_file.digest` output, but use `:push_test_tag_file` instead. That file should contain a pusher command that includes the tag of the container. You should make sure that tag is correct.

So try to do this on your command line and you will see:

```
bazel build //tests/docker:push_test_tag_file
cat bazel-bin/tests/docker/push_test_tag_file
```
Then in the `content` you want to validate in the `file_test` rule, you need to check that file containers a regex that has the tag in it.

**Comment by excavador** [comment]
@xingao267 thank you so much, now got it!
I knew about the executable script file but did not imagine to use it for verification.

Tests added, rebased to latest master, seems like now I fixed everything

**Comment by xingao267** [comment]
Thanks! Looks good!

**Pull request merged**